File: tests/refs/bst_validate.py
"""Reference: validate a binary search tree.

The fixture encodes a binary tree as a level-order list using `None` for
missing children (LeetCode-style). The driver reconstructs the tree and runs
the recursive bounds check.

A strict BST requires `lo < node.val < hi`.
"""

from __future__ import annotations
from dataclasses import dataclass
from math import inf
from typing import List, Optional, Any


@dataclass
class _Node:
    val: int
    left: Optional["_Node"] = None
    right: Optional["_Node"] = None


def _build(values: List[Any]) -> Optional[_Node]:
    if not values or values[0] is None:
        return None
    root = _Node(values[0])
    queue: List[_Node] = [root]
    i = 1
    while queue and i < len(values):
        node = queue.pop(0)
        if i < len(values):
            v = values[i]
            if v is not None:
                node.left = _Node(v)
                queue.append(node.left)
            i += 1
        if i < len(values):
            v = values[i]
            if v is not None:
                node.right = _Node(v)
                queue.append(node.right)
            i += 1
    return root


def isValidBST(values: List[Any]) -> bool:
    root = _build(values)

    def check(n: Optional[_Node], lo: float, hi: float) -> bool:
        if n is None:
            return True
        if not (lo < n.val < hi):
            return False
        return check(n.left, lo, n.val) and check(n.right, n.val, hi)

    return check(root, -inf, inf)

File: tests/refs/BstValidate.java
/*
 * Reference Java implementation for tests/cases/bst_validate.json.
 *
 * Tree is encoded as a LeetCode-style level-order list with `null` for
 * missing children (passed in as a List<Long> with possible nulls). The
 * driver reconstructs the tree and validates strict BST bounds.
 */
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.List;

public class BstValidate {
    private static class Node {
        long val;
        Node left, right;
        Node(long v) { this.val = v; }
    }

    private static Node build(List<Long> values) {
        if (values.isEmpty() || values.get(0) == null) return null;
        Node root = new Node(values.get(0));
        Deque<Node> q = new ArrayDeque<>();
        q.offer(root);
        int i = 1;
        while (!q.isEmpty() && i < values.size()) {
            Node node = q.poll();
            if (i < values.size()) {
                Long v = values.get(i++);
                if (v != null) { node.left = new Node(v); q.offer(node.left); }
            }
            if (i < values.size()) {
                Long v = values.get(i++);
                if (v != null) { node.right = new Node(v); q.offer(node.right); }
            }
        }
        return root;
    }

    private static boolean check(Node n, long lo, long hi, boolean hasLo, boolean hasHi) {
        if (n == null) return true;
        if (hasLo && n.val <= lo) return false;
        if (hasHi && n.val >= hi) return false;
        return check(n.left, lo, n.val, hasLo, true)
            && check(n.right, n.val, hi, true, hasHi);
    }

    public static boolean isValidBST(List<Long> values) {
        Node root = build(values);
        return check(root, 0, 0, false, false);
    }
}
